### ISSUE TITLE
Disconnect with "no recent network activity" because of lost MTU probe packet

### ISSUE THREAD

**Issue by omgold** [lucas-clemente/quic-go#3409]
I discovered when using Syncthing in quic mode, in some cases endless loops of connect, reconnect., with the debug output giving the reason

`client Destroying connection: timeout: no recent network activity`

according to tcpdump the only packets are dropped are MTU probe f

```
2022/05/12 10:09:28 client -> Sending packet 14 (1439 bytes) for connection 621bb0fa28e866c1aec42f56f203, 1-RTT
2022/05/12 10:09:28 client 	Short Header{DestConnectionID: 8ec71731, PacketNumber: 14, PacketNumberLen: 2, KeyPhase: 0}
2022/05/12 10:09:28 client 	-> &wire.PingFrame{}
```

The MTU here is correct discovered as 1427, so this (too large) packet does not reach the server.

Digging into the code, the way I understand it, connections are dropped when no traffic is received for 30s. To prevent this, keepalive probes are send after 15s, in case there are no unanswered "ack eliciting packets":

```
// run the connection main loop
func (s *connection) run() error {
        ....
runLoop:
	for {
                ...
		if keepAliveTime := s.nextKeepAliveTime(); !keepAliveTime.IsZero() && !now.Before(keepAliveTime) {
			// send a PING frame since there is no activity in the connection
			s.logger.Debugf("Sending a keep-alive PING to keep the connection alive.")
			s.framer.QueueControlFrame(&wire.PingFrame{})
			s.keepAlivePingSent = true
		}
                ...
        }
        ...
```

With `nextKeepAliveTime` computed like this:

```
func (s *connection) nextKeepAliveTime() time.Time {
	if !s.config.KeepAlive || s.keepAlivePingSent || !s.firstAckElicitingPacketAfterIdleSentTime.IsZero() {
		return time.Time{}
	}
	return s.lastPacketReceivedTime.Add(s.keepAliveInterval)
}
```

Now when a MTU probe packet is send:

```
func (s *connection) sendPacket() (bool, error) {
        ...
	if !s.config.DisablePathMTUDiscovery && s.mtuDiscoverer.ShouldSendProbe(now) {
		packet, err := s.packer.PackMTUProbePacket(s.mtuDiscoverer.GetPing())
		if err != nil {
			return false, err
		}
		s.sendPackedPacket(packet, now)
		return true, nil
	}
        ...
}
```

The `sendPacketPacket` function updates the `firstAckElicitingPacketAfterIdleSentTime`, because `packet.IsAckEliciting()` return true for a `PackMTUProbePacket`:

```
func (s *connection) sendPackedPacket(packet *packedPacket, now time.Time) {
	if s.firstAckElicitingPacketAfterIdleSentTime.IsZero() && packet.IsAckEliciting() {
		s.firstAckElicitingPacketAfterIdleSentTime = now
	}
        ....
}
```

In short, the reason is that it is assumed that a MTU probe packet is always answered (and thus no keepalive is necessary), but this is of course not the case by the nature of MTU probes.

**Comment by marten-seemann** [comment]
Nice work, and excellent writeup! Thank you!

The problem would be solved if we didn't set `firstAckElicitingPacketAfterIdleSentTime` for MTU probe packets, right?

**Comment by marten-seemann** [comment]
I'm wondering if there are any unexpected consequences regarding the [idle timeout](https://datatracker.ietf.org/doc/html/rfc9000#section-10.1):
> An endpoint restarts its idle timer when a packet from its peer is received and processed successfully.  An endpoint also restarts its idle timer when sending an ack-eliciting packet if no other ack-eliciting packets have been sent since last receiving and processing a packet.  Restarting this timer when sending a packet ensures that connections are not closed after new activity is initiated.

The peer receiving the probe packet will interpret that packet as resetting the idle timeout timer, but (with the change suggested) the peer sending the packet will not.

---

Maybe the DPLPMTUD implemetation is too aggressive in the first place? Maybe it would make sense to not send MTU probe packets when we'd be idle otherwise?

**Comment by omgold** [comment]
> The problem would be solved if we didn't set `firstAckElicitingPacketAfterIdleSentTime` for MTU probe packets, right?

I did't test it, but I would very much believe so.

**Comment by omgold** [comment]
> I'm wondering if there are any unexpected consequences regarding the [idle timeout](https://datatracker.ietf.org/doc/html/rfc9000#section-10.1):
> 
> > An endpoint restarts its idle timer when a packet from its peer is received and processed successfully.  An endpoint also restarts its idle timer when sending an ack-eliciting packet if no other ack-eliciting packets have been sent since last receiving and processing a packet.  Restarting this timer when sending a packet ensures that connections are not closed after new activity is initiated.
> 
> The peer receiving the probe packet will interpret that packet as resetting the idle timeout timer, but (with the change suggested) the peer sending the packet will not.
> 
> Maybe the DPLPMTUD implemetation is too aggressive in the first place? Maybe it would make sense to not send MTU probe packets when we'd be idle otherwise?

Another question is, whether a single dropped packet should immediately result in a connection drop, and not in a couple of resends first. I'm not familiar enough with the quic protocol, though, to understand whether that would be allowed within the standard.

**Comment by marten-seemann** [comment]
> Another question is, whether a single dropped packet should immediately result in a connection drop, and not in a couple of resends first. I'm not familiar enough with the quic protocol, though, to understand whether that would be allowed within the standard.

That's not what's happening. QUIC connections use an timer to silently close connections that are "idle" (i.e. no packets sent / received for a long time). Normal packets are retransmitted until they're acknowledged by the peer - but this doesn't apply to Path MTU Discovery packets, which is where the mismatch here comes from.

**Comment by marten-seemann** [comment]
Started a discussion on the QUIC mailing list: https://mailarchive.ietf.org/arch/msg/quic/HSo9yB_6wn1dCHtO7EDyYgZbNPA/
